[1/n] IPv6 support: address <-> host:port conversion

Yicheng-Lu-llll opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

This is a continuation of the work in [#54662 (comment)](https://github.com/ray-project/ray/pull/54662#issuecomment-3130750868).

Please refer to that discussion for more context
<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

Yicheng-Lu-llll commented on src/ray/common/network_util.cc:
```diff
@@ -28,3 +33,50 @@ bool CheckPortFree(int port) {
+  }
+}
+
+std::vector<std::string> ParseAddress(const std::string &address) {
```
Here, we are not aiming to provide strict address validation, but rather to mimic the behavior of `absl::StrSplit(ipv4_address, ':')`.

For example, in `absl::StrSplit(ipv4_address, ':')`:

If the `ipv4_address` contains a colon (:), it is assumed to be in the ip:port format and is split accordingly.

If the `ipv4_address` does not contain a colon, it returns `[ipv4_address]`, treating it as an address without a specified port.

Similarly, ParseAddress provides the same behavior for IPv6:

If `[2001:db8::1]:8000` is provided, it returns `["2001:db8::1", "8000"]`.

If `2001:db8::1 `or even `::1` is provided, it simply returns `["2001:db8::1"]` or `["::1"]`

aslonnie reviewed (commented):
this is changing a lot of files.. can we split this into smaller PRs and roll out this more incrementally?

jjyao commented on src/ray/util/util.cc:
```diff
@@ -225,3 +228,49 @@ std::optional<std::chrono::steady_clock::time_point> ToTimeoutPoint(int64_t time
+
+std::vector<std::string> ParseAddress(const std::string &address) {
+  size_t pos = address.find_last_of(":");
+  if (pos == std::string::npos) {
+    return {address};
+  }
```
When will we call this function without port?

Yicheng-Lu-llll replied:
This is aim to mimic the behavior of split. For example, we currently have code where the input can be either a plain IP address or an IP address with a port (ip:port).
https://github.com/ray-project/ray/blob/bb508648dfccc1777950e7beef5a38243cd3becd/python/ray/_private/services.py#L592-L612

jjyao commented on src/ray/protobuf/object_manager.proto:
```diff
@@ -50,12 +50,9 @@ message FreeObjectsRequest {
-}
-message FreeObjectsReply {
-}
+message PushReply {}
```
To make the review easier, could you create a separate PR for this kind of pure format changes?

jjyao reviewed (commented):
Thanks for working on this!!

Yicheng-Lu-llll commented:
I have:
- Moved `BuildAddress` and `ParseAddress` to network_util.h/.cc
- Removed formatting changes to the proto file
- Made `resolve_ip_for_localhost` accept only IP addresses

will soon separate the PR into C++ and Python changes and recheck every changes.

Yicheng-Lu-llll committed:
IPv6 support: address <-> host:port conversion

Signed-off-by: Yicheng-Lu-llll <luyc58576@gmail.com>
